saleor/order/utils.py

```diff
@@ -909,123 +888,124 @@ def update_discount_for_order_line(
-                "base_unit_price_amount",
-                "undiscounted_unit_price_gross_amount",
-                "undiscounted_unit_price_net_amount",
-                "undiscounted_total_price_gross_amount",
-                "undiscounted_total_price_net_amount",
-            ]
-        )
+        update_unit_discount_data_on_order_line(order_line, line_discounts)
+        fields_to_update = [
+            "unit_discount_value",
+            "unit_discount_amount",
+            "unit_discount_type",
+            "unit_discount_reason",
+            "base_unit_price_amount",
+        ]
+        order_line.save(update_fields=fields_to_update)
 
-    # Save lines before calculating the taxes as some plugin can fetch all order data
-    # from db
-    order_line.save(update_fields=fields_to_update)
 
-    _update_manual_order_line_discount_object(
-        value, value_type, reason, order_line, order.currency
-    )
+def _remove_invalid_discounts_for_adding_manual(
+    order_line_discounts: list[OrderLineDiscount],
+):
+    """Remove all line discounts except the single manual line discount."""
+    discount_to_delete = []
+    current_manual_discount = None
+    for discount in order_line_discounts:
+        if discount.type == DiscountType.MANUAL and not current_manual_discount:
+            current_manual_discount = discount
+        else:
+            discount_to_delete.append(discount)
 
+    if discount_to_delete:
+        for discount in discount_to_delete:
+            order_line_discounts.remove(discount)
+        OrderLineDiscount.objects.filter(
+            id__in=[discount.id for discount in discount_to_delete]
+        ).delete()
 
-def _update_manual_order_line_discount_object(
-    value, value_type, reason, order_line, currency
-):
+
+def _get_order_line_discount_to_update(
+    order_line_discounts: list[OrderLineDiscount],
+) -> OrderLineDiscount | None:
     discount_to_update = None
-    discount_to_delete_ids = []
-    discounts = order_line.discounts.all()
-    for discount in discounts:
-        if discount.type == DiscountType.MANUAL and not discount_to_update:
+    for discount in order_line_discounts:
+        if discount.type == DiscountType.MANUAL:
             discount_to_update = discount
-        else:
-            discount_to_delete_ids.append(discount.pk)
+            break
+    return discount_to_update
 
-    if discount_to_delete_ids:
-        OrderLineDiscount.objects.filter(id__in=discount_to_delete_ids).delete()
 
-    amount_value = quantize_price(
-        order_line.unit_discount.amount * order_line.quantity, currency
-    )
-    if not discount_to_update:
-        order_line.discounts.create(
-            type=DiscountType.MANUAL,
-            value_type=value_type,
-            value=value,
-            amount_value=amount_value,
-            currency=currency,
-            reason=reason,
-            unique_type=DiscountType.MANUAL,
+def _create_or_update_manual_order_line_discount_object(
```

Comment:
suggestion (non-blocking): seems like this function is not creating the line discount anymore, so the name should be changed